File: pkg/connectors/twitter.go
package connectors

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/jschell12/replicateme/pkg/corpus"
)

// TwitterImportOptions controls Twitter/X archive import.
type TwitterImportOptions struct {
	File       string // path to ZIP or extracted archive directory
	Since      *time.Time
	IncludeDMs *bool // default true
}

type tweetEntry struct {
	Tweet struct {
		IDStr                  string `json:"id_str"`
		FullText               string `json:"full_text"`
		CreatedAt              string `json:"created_at"`
		InReplyToStatusIDStr   string `json:"in_reply_to_status_id_str,omitempty"`
		InReplyToScreenName    string `json:"in_reply_to_screen_name,omitempty"`
	} `json:"tweet"`
}

type dmConversationWrapper struct {
	DMConversation struct {
		Messages []struct {
			MessageCreate *struct {
				ID          string `json:"id"`
				Text        string `json:"text"`
				CreatedAt   string `json:"createdAt"`
				SenderID    string `json:"senderId"`
				RecipientID string `json:"recipientId"`
			} `json:"messageCreate,omitempty"`
		} `json:"messages"`
	} `json:"dmConversation"`
}

// ImportTwitter reads tweets and DMs from a Twitter/X data archive.
func ImportTwitter(opts TwitterImportOptions) ([]corpus.RawMessage, error) {
	dir := opts.File

	if strings.HasSuffix(opts.File, ".zip") {
		dir = "/tmp/replicateme-twitter-export"
		if err := extractZip(opts.File, dir); err != nil {
			return nil, fmt.Errorf("extract twitter zip: %w", err)
		}
	}

	if _, err := os.Stat(dir); err != nil {
		return nil, fmt.Errorf("Twitter archive not found at %s", dir)
	}

	dataDir := dir
	if _, err := os.Stat(filepath.Join(dir, "data")); err == nil {
		dataDir = filepath.Join(dir, "data")
	}

	var messages []corpus.RawMessage

	// import tweets
	tweetsFile := findFile(dataDir, []string{"tweets.js", "tweet.js"})
	if tweetsFile != "" {
		var tweets []tweetEntry
		if err := parseTwitterJS(tweetsFile, &tweets); err == nil {
			for _, entry := range tweets {
				t := entry.Tweet
				if t.FullText == "" {
					continue
				}
				ts, err := time.Parse("Mon Jan 02 15:04:05 -0700 2006", t.CreatedAt)
				if err != nil {
					continue
				}
				if opts.Since != nil && ts.Before(*opts.Since) {
					continue
				}

				messages = append(messages, corpus.RawMessage{
					ID:         fmt.Sprintf("twitter-%s", t.IDStr),
					Text:       cleanTweetText(t.FullText),
					Platform:   corpus.Twitter,
					Timestamp:  ts.UTC(),
					IsFromUser: true,
					Metadata: map[string]any{
						"isReply": t.InReplyToStatusIDStr != "",
						"replyTo": t.InReplyToScreenName,
					},
				})
			}
		}
	}

	// import DMs
	includeDMs := opts.IncludeDMs == nil || *opts.IncludeDMs
	if includeDMs {
		dmsFile := findFile(dataDir, []string{"direct-messages.js", "direct_messages.js"})
		if dmsFile != "" {
			var dms []dmConversationWrapper
			if err := parseTwitterJS(dmsFile, &dms); err == nil {
				for _, conv := range dms {
					for _, msg := range conv.DMConversation.Messages {
						dm := msg.MessageCreate
						if dm == nil || dm.Text == "" {
							continue
						}
						createdAtMs, err := strconv.ParseInt(dm.CreatedAt, 10, 64)
						if err != nil {
							continue
						}
						ts := time.UnixMilli(createdAtMs).UTC()
						if opts.Since != nil && ts.Before(*opts.Since) {
							continue
						}

						messages = append(messages, corpus.RawMessage{
							ID:         fmt.Sprintf("twitter-dm-%s", dm.ID),
							Text:       dm.Text,
							Platform:   corpus.Twitter,
							Timestamp:  ts,
							IsFromUser: true,
							Metadata: map[string]any{
								"isDM":        true,
								"senderId":    dm.SenderID,
								"recipientId": dm.RecipientID,
							},
						})
					}
				}
			}
		}
	}

	sort.Slice(messages, func(i, j int) bool {
		return messages[i].Timestamp.Before(messages[j].Timestamp)
	})
	return messages, nil
}

func parseTwitterJS(filePath string, target any) error {
	content, err := os.ReadFile(filePath)
	if err != nil {
		return err
	}
	s := string(content)
	// strip "window.YTD.xxx.part0 = " prefix
	if idx := strings.Index(s, "="); idx != -1 {
		s = strings.TrimSpace(s[idx+1:])
	}
	// strip trailing semicolon
	s = strings.TrimRight(s, ";")
	return json.Unmarshal([]byte(s), target)
}

func findFile(dir string, names []string) string {
	for _, name := range names {
		path := filepath.Join(dir, name)
		if _, err := os.Stat(path); err == nil {
			return path
		}
	}
	return ""
}

var tcoURLRe = regexp.MustCompile(`https?://t\.co/\w+`)

func cleanTweetText(text string) string {
	cleaned := tcoURLRe.ReplaceAllString(text, "")
	cleaned = strings.TrimSpace(cleaned)
	cleaned = strings.ReplaceAll(cleaned, "&amp;", "&")
	cleaned = strings.ReplaceAll(cleaned, "&lt;", "<")
	cleaned = strings.ReplaceAll(cleaned, "&gt;", ">")
	cleaned = strings.ReplaceAll(cleaned, "&quot;", `"`)
	return cleaned
}

File: src/connectors/twitter.ts
import { readFileSync, existsSync } from "fs";
import { join } from "path";
import { execSync } from "child_process";
import type { RawMessage } from "../corpus/types.js";

// Twitter/X data archive format:
// ZIP containing data/ directory with JS files.
// tweets.js: window.YTD.tweet.part0 = [ { tweet: { full_text, created_at, id_str, ... } } ]
// direct-messages.js: similar structure

interface TweetEntry {
  tweet: {
    id_str: string;
    full_text: string;
    created_at: string;
    in_reply_to_status_id_str?: string;
    in_reply_to_screen_name?: string;
  };
}

export interface TwitterImportOptions {
  file: string; // path to ZIP or extracted archive directory
  since?: Date;
  includeDMs?: boolean;
}

export function importTwitter(opts: TwitterImportOptions): RawMessage[] {
  let dir = opts.file;

  if (opts.file.endsWith(".zip")) {
    dir = "/tmp/replicateme-twitter-export";
    execSync(`rm -rf ${dir} && mkdir -p ${dir} && unzip -q -o "${opts.file}" -d ${dir}`);
  }

  if (!existsSync(dir)) {
    throw new Error(`Twitter archive not found at ${dir}`);
  }

  // twitter archive may have a nested directory
  const dataDir = existsSync(join(dir, "data")) ? join(dir, "data") : dir;

  const messages: RawMessage[] = [];

  // import tweets
  const tweetsFile = findFile(dataDir, ["tweets.js", "tweet.js"]);
  if (tweetsFile) {
    const tweets = parseTwitterJS<TweetEntry[]>(tweetsFile);
    for (const entry of tweets) {
      const t = entry.tweet;
      if (!t.full_text) continue;

      const timestamp = new Date(t.created_at);
      if (opts.since && timestamp < opts.since) continue;

      messages.push({
        id: `twitter-${t.id_str}`,
        text: cleanTweetText(t.full_text),
        platform: "twitter",
        timestamp,
        isFromUser: true,
        metadata: {
          isReply: !!t.in_reply_to_status_id_str,
          replyTo: t.in_reply_to_screen_name,
        },
      });
    }
  }

  // import DMs if requested
  if (opts.includeDMs !== false) {
    const dmsFile = findFile(dataDir, [
      "direct-messages.js",
      "direct_messages.js",
    ]);
    if (dmsFile) {
      const dms = parseTwitterJS<Array<{ dmConversation: DMConversation }>>(dmsFile);
      for (const conv of dms) {
        for (const msg of conv.dmConversation?.messages ?? []) {
          const dm = msg.messageCreate;
          if (!dm?.text) continue;

          const timestamp = new Date(parseInt(dm.createdAt));
          if (opts.since && timestamp < opts.since) continue;

          // we can't reliably tell which is "from me" without the user ID,
          // but twitter archive DMs include senderId - mark all for now,
          // user can filter later
          messages.push({
            id: `twitter-dm-${dm.id}`,
            text: dm.text,
            platform: "twitter",
            timestamp,
            isFromUser: true,
            metadata: {
              isDM: true,
              senderId: dm.senderId,
              recipientId: dm.recipientId,
            },
          });
        }
      }
    }
  }

  messages.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
  return messages;
}

interface DMConversation {
  messages: Array<{
    messageCreate?: {
      id: string;
      text: string;
      createdAt: string;
      senderId: string;
      recipientId: string;
    };
  }>;
}

function parseTwitterJS<T>(filePath: string): T {
  let content = readFileSync(filePath, "utf-8");
  // strip the "window.YTD.xxx.part0 = " prefix
  const eqIdx = content.indexOf("=");
  if (eqIdx !== -1) {
    content = content.substring(eqIdx + 1).trim();
  }
  // strip trailing semicolon
  if (content.endsWith(";")) {
    content = content.slice(0, -1);
  }
  return JSON.parse(content);
}

function findFile(dir: string, names: string[]): string | null {
  for (const name of names) {
    const path = join(dir, name);
    if (existsSync(path)) return path;
  }
  return null;
}

function cleanTweetText(text: string): string {
  // remove t.co URLs
  let cleaned = text.replace(/https?:\/\/t\.co\/\w+/g, "").trim();
  // decode HTML entities
  cleaned = cleaned
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"');
  return cleaned;
}
